proposals/0000-fine-grained-undecidable.md

```diff
@@ -0,0 +1,198 @@
+
+```haskell
+%NoTerminationCheck ; instance {-# OVERLAPS #-} ...
+```
+
+For that reason, this proposal also regularizes the syntax, turning those pragmas into modifiers `%Overlapping`, `%Overlappable`, and `%Overlaps`.
+
+## Motivation
+
+When working with type classes and families, sometimes we write definitions which lay outside the [Paterson conditions](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/jfp06.pdf). Roughly speaking, that means that GHC is not able to prove the termination of the resolution / constraint solving process, and thus rejects the definition. One particular case is when the context of an instance mentions a variable more often than its head.
+
+```haskell
+data Pair a = Pair (a, a)
+instance Show (a, a) => Show (Pair a) where ...
+```
+
+As GHC's error message point out, one can escape those checks by using `UndecidableInstances`.
+
+```
+<interactive>: error:
+    • Variable ‘a’ occurs more often
+        in the constraint ‘Show (a, a)’
+        than in the instance head ‘Show (Pair a)’
+      (Use UndecidableInstances to permit this)
+```
+
+However, this is quite a drastic measure. In order to accept _one_ instance, we need to disable the checks for _every_ instance defined in the module. For example, now looping instances are allowed without warning.
+
+```haskell
+instance Show (Pair a) => Show (Pair a)
+```
+
+### The many faces of `Undecidable`
+
+As it turns out, GHC puts a lot of tweak under the `UndecidableSuperClasses` and `UndecidableInstances` extensions. A few of them are related to termination proper:
+
+- `UndecidableSuperClasses` omits the cycle check for all class declarations in a module. This allows for arbitrary predicates in a superclass constraint.
+- `UndecidableInstances` omits the termination check for all local quantified constraints in a module.
+- `UndecidableInstances` skips the termination check for class instances.
+- `UndecidableInstances` skips the termination check for type family equations (both in open type family instances and closed type family equations).
+
+There are also tweaks to the coverage conditions for coverage / injectivity in type families and functional dependencies.
+
+- `UndecidableInstances` allows a liberalization of the check for injective type family consistency.
+- `UndecidableInstances` allows an instance of a class with functional dependencies to use the liberal coverage condition instead of the coverage condition.
+
+## Proposed Change Specification
+
+The modifier syntax is taken from this [approved proposal](https://github.com/ghc-proposals/ghc-proposals/pull/370), with the additional [change](https://github.com/ghc-proposals/ghc-proposals/pull/392) of modifiers appearing _before_ the modified thing. Since modifiers ought to be types, we introduce the following declarationg in the `GHC.Generics` module.
```

Comment:
Why GHC.Generics in particular?

Reply:
Oops, this was meant to be `GHC.Modifiers` instead.